### PR TITLE
HIP-584: Extract util classes needed for precompile logic

### LINKED ISSUE #5653: HIP-584: Extract remaining encoding and decoding utils needed for precompiles
### Problem

In order to provide implementations for all precompiles, we need to copy logic related to encoding and decoding as well as some constants from `hedera-services` to `hedera-mirror-node`.

### Solution

Copy logic from:

- EncodingFacade
- DecodingFacade
- ParsingConstants

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by ivan-02**
**Description**:

Copied logic related to encoding and decoding as well as some constants from hedera-services to hedera-mirror-node. These classes are needed for the implementation of the precompiles.  

**Related issue(s)**:

Fixes #5653

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/utils/ByteStringUtils.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import edu.umd.cs.findbugs.annotations.NonNull;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.function.Function;
+
+/** Protobuf related utilities shared by client and server. */
+public final class ByteStringUtils {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/hapi/utils/ByteStringUtils.java:32]: I think we can get rid of this class

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+            final boolean isApproval = (transfer.size() > 2) && (boolean) transfer.get(2);
+            addSignedAdjustment(fungibleTransfers, tokenType, accountID, amount, isApproval);
+        }
+        return fungibleTransfers;
+    }
+
+    @NonNull
+    public static AccountID generateAccountIDWithAliasCalculatedFrom(final AccountID accountID) {
+        return AccountID.newBuilder()
+                .setAlias(ByteStringUtils.wrapUnsafely(EntityIdUtils.asEvmAddress(accountID)))
```
[the remark this answers is not in the document]
Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java:222]: I suggest using `UnsafeByteOperations.unsafeWrap(EntityIdUtils.asEvmAddress(accountID))` , as ByteString.wrap is package private field

File: hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityIdUtils.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                    .setShardNum(parts[0])
+                    .setRealmNum(parts[1])
+                    .setAccountNum(parts[2])
+                    .build();
+        } catch (final NumberFormatException | ArrayIndexOutOfBoundsException e) {
+            throw new IllegalArgumentException(String.format("Argument 'literal=%s' is not an account", literal), e);
+        }
+    }
+
+    private static long[] parseLongTriple(final String dotDelimited) {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityIdUtils.java:49]: We can delete this method

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/FungibleTokenTransfer.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package com.hedera.services.store.contracts.precompile;
+
```
Comment (edwin-greene) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/FungibleTokenTransfer.java:2]: A few of these new classes need licenses.

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/utils/ByteStringUtils.java (hunk 1 above)
Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/hapi/utils/ByteStringUtils.java]: Done

File: hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityIdUtils.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+public final class EntityIdUtils {
+
+    private static final String CANNOT_PARSE_PREFIX = "Cannot parse '";
+
+    private EntityIdUtils() {
+        throw new UnsupportedOperationException("Utility Class");
+    }
+
+    public static AccountID parseAccount(final String literal) {
```
[the remark this answers is not in the document]
Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityIdUtils.java]: Done

File: hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityIdUtils.java (hunk 3 above)
Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityIdUtils.java]: Done

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/FungibleTokenTransfer.java (hunk 4 above)
Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/FungibleTokenTransfer.java:17]: Done

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    public static TokenExpiryWrapper decodeTokenExpiry(
+            @NonNull final Tuple expiryTuple, final UnaryOperator<byte[]> aliasResolver) {
+        final var second = (long) expiryTuple.get(0);
+        final var autoRenewAccount = convertLeftPaddedAddressToAccountId(expiryTuple.get(1), aliasResolver);
+        final var autoRenewPeriod = (long) expiryTuple.get(2);
+        return new TokenExpiryWrapper(
+                second, autoRenewAccount.getAccountNum() == 0 ? null : autoRenewAccount, autoRenewPeriod);
+    }
+
+    public static Tuple decodeFunctionCall(
```
[the remark this answers is not in the document]
Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java]: Done

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/utils/contracts/ParsingConstants.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.services.hapi.utils.contracts;
+
+import com.esaulpaugh.headlong.abi.TupleType;
+
+public final class ParsingConstants {
```
[the remark this answers is not in the document]
Comment (ivan-02) [reply hedera-mirror-web3/src/main/java/com/hedera/services/hapi/utils/contracts/ParsingConstants.java:22]: Done

File: hedera-mirror-web3/src/test/java/com/hedera/services/store/contracts/precompile/codec/EncodingFacadeTest.java [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+    void createsExpectedMintFailureResult() {
+        assertEquals(MINT_FAILURE_FROM_INVALID_TOKEN_ID, subject.encodeMintFailure(INVALID_TOKEN_ID));
+    }
+
+    @Test
+    void createsExpectedBurnFailureResult() {
+        assertEquals(BURN_FAILURE_FROM_TREASURY_NOT_OWNER, subject.encodeBurnFailure(TREASURY_MUST_OWN_BURNED_NFT));
+    }
+
+    private com.esaulpaugh.headlong.abi.Address convertBesuAddressToHeadlongAddress(
```
[the remark this answers is not in the document]
Comment (ivan-02) [reply hedera-mirror-web3/src/test/java/com/hedera/services/store/contracts/precompile/codec/EncodingFacadeTest.java]: Done

File: hedera-mirror-web3/src/test/java/com/hedera/services/store/contracts/precompile/codec/TokenExpiryWrapperTest.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+                .setRealmNum(nativeParts[1])
+                .setAccountNum(nativeParts[2])
+                .build();
+    }
+
+    //copied from IdUtils
+    static long[] asDotDelimitedLongArray(String s) {
+        String[] parts = s.split("[.]");
+        return Stream.of(parts).mapToLong(Long::valueOf).toArray();
+    }
```
[the remark this answers is not in the document]
Comment (ivan-02) [reply hedera-mirror-web3/src/test/java/com/hedera/services/store/contracts/precompile/codec/TokenExpiryWrapperTest.java]: Done

**Review by steven-sheehy — APPROVED** [review]

**Pull request merged**